packages/tx/src/transaction.ts

```diff
@@ -21,340 +21,373 @@ const N_DIV_2 = new BN('7fffffffffffffffffffffffffffffff5d576e7357a4501ddfe92f46
+      bnToRlp(this.v),
+      bnToRlp(this.r),
+      bnToRlp(this.s),
+    ]
+
+    return rlphash(values)
+  }
 
-    // create hash
-    return rlphash(items)
+  getMessageToSign() {
+    return this._getMessageToSign(this._unsignedTxImplementsEIP155())
+  }
+
+  getMessageToVerifySignature() {
+    return this._getMessageToSign(this._signedTxImplementsEIP155())
   }
 
   /**
-   * returns chain ID
+   * Returns chain ID
    */
   getChainId(): number {
-    return this._common.chainId()
+    return this.common.chainId()
   }
 
   /**
-   * returns the sender's address
+   * Returns the sender's address
    */
-  getSenderAddress(): Buffer {
-    if (this._from) {
-      return this._from
-    }
-    const pubkey = this.getSenderPublicKey()
-    this._from = publicToAddress(pubkey)
-    return this._from
+  getSenderAddress(): Address {
+    return new Address(publicToAddress(this.getSenderPublicKey()))
   }
 
   /**
-   * returns the public key of the sender
+   * Returns the public key of the sender
    */
   getSenderPublicKey(): Buffer {
-    if (!this.verifySignature()) {
-      throw new Error('Invalid Signature')
+    const msgHash = this.getMessageToVerifySignature()
+
+    // All transaction signatures whose s-value is greater than secp256k1n/2 are considered invalid.
+    if (this.common.gteHardfork('homestead') && this.s && this.s.gt(N_DIV_2)) {
+      throw new Error(
+        'Invalid Signature: s-values greater than secp256k1n/2 are considered invalid',
+      )
     }
 
-    // If the signature was verified successfully the _senderPubKey field is defined
-    return this._senderPubKey!
+    try {
+      return ecrecover(
+        msgHash,
+        this.v!.toNumber(),
+        bnToRlp(this.r),
+        bnToRlp(this.s),
+        this._signedTxImplementsEIP155() ? this.getChainId() : undefined,
+      )
+    } catch (e) {
+      throw new Error('Invalid Signature')
+    }
   }
 
   /**
    * Determines if the signature is valid
    */
   verifySignature(): boolean {
-    const msgHash = this.hash(false)
-    // All transaction signatures whose s-value is greater than secp256k1n/2 are considered invalid.
-    if (this._common.gteHardfork('homestead') && new BN(this.s).cmp(N_DIV_2) === 1) {
-      return false
-    }
-
     try {
-      const v = bufferToInt(this.v)
-      const useChainIdWhileRecoveringPubKey =
-        v >= this.getChainId() * 2 + 35 && this._common.gteHardfork('spuriousDragon')
-      this._senderPubKey = ecrecover(
-        msgHash,
-        v,
-        this.r,
-        this.s,
-        useChainIdWhileRecoveringPubKey ? this.getChainId() : undefined,
-      )
+      return unpadBuffer(this.getSenderPublicKey()).length !== 0
     } catch (e) {
       return false
     }
-
-    return !!this._senderPubKey
   }
 
   /**
-   * sign a transaction with a given private key
+   * Sign a transaction with a given private key
    * @param privateKey - Must be 32 bytes in length
    */
   sign(privateKey: Buffer) {
```

Comment:
I think we should point this out more clearly in the code docs that this only gives a signed transaction if assigned to a new variable, otherwise this is too dangerous that people assume their transaction to be signed when just using `tx.sign()` (and not further thinking about immutability, which is in fact rather likely than unlikely I would assume).